### PR TITLE
Support lifetime elision in HIR

### PR DESCRIPTION
**Pull request by QnnOkabayashi**
This is the third PR towards having a type universe (https://github.com/rust-diplomat/diplomat/issues/204).

# This PR:
* Adds lifetime elision inference in the HIR using the elision rules according to the [Rustonomicon](https://doc.rust-lang.org/nomicon/lifetime-elision.html), which says:
  > Elision rules are as follows:
  > * Each elided lifetime in input position becomes a distinct lifetime parameter.
  > * If there is exactly one input lifetime position (elided or not), that lifetime is assigned to all elided output lifetimes.
  > * If there are multiple input lifetime positions, but one of them is &self or &mut self, the lifetime of self is assigned to all elided output lifetimes.
  > * Otherwise, it is an error to elide an output lifetime.
* Makes progress towards https://github.com/rust-diplomat/diplomat/issues/160

# Remaining work
There are some weird examples where rustc can figure out elision, but we can't. This is caused by the fact that when searching for a lifetime in the input that can be tied to elided lifetimes in the output, rustc doesn't look at the lifetimes in the generics of the `self` type, _and_ it also doesn't look at any parameter that has the type `Self`. We can't figure this out because the AST eagerly substitutes that actual type in for `Self`, so we don't know what was and wasn't `Self` by the time we lower and do elision inferece. This means changes to `ast` are required. Here's an example of where this breaks the elision inference implementation in this PR:
```rust
struct Foo<'x>(&'x str);

impl<'x> Foo<'x> {
    // rustc doesn't see the `'x` in the type of `self` or in `a: Self`,
    // so it thinks the only lifetime is on `b: Foo<'x>`

    // Like rustc, we also skip the `'x` in `self`, but we don't skip the `'x` in `a: Self`,
    // so we think there are multiple lifetimes and that the elision is ambiguous.
    fn bar(self: Foo<'x>, a: Self, b: Foo<'x>) -> &str {
        b.0
    }
}
```

# Questions
I made this comment awhile ago about not needing to detect implicit bounds...

> For JS, I realized this isn't actually that much of an issue because even though `LifetimeEnv` doesn't get that information, it's impossible for `Foo<'b>` to drop before `'a` expires _if we return something with `x`_. This is because opaques also hold onto their owner if borrowed.
> We would still need to add a validity check because the C++ backend is still affected by this.
> _Originally posted by @QnnOkabayashi in https://github.com/rust-diplomat/diplomat/issues/198#issuecomment-1179190150_

... and I think I was right about the JS, but I was wrong about C++ not working since you can just follow the same reasoning as JS. That is, if B borrows A, and C borrows B, the docs will tells C that it can't outlive B, and somewhere else it will say that B can't outlive A. So if the programmer is following those invariants, then it's impossible for that implicit lifetime bound to be broken.

### CODE REVIEW

**Comment by sffc** [comment]
Thanks for the PR! I'll defer to @Manishearth on the details...

**Comment by QnnOkabayashi** [comment]
I made another branch off of this branch that adds the `TypeName::SelfType(PathType)` variant and makes everything work in all the backends, so I'm just waiting for this to merge before I can send that one.

File: core/src/hir/lowering.rs
Comment (Manishearth) [inline core/src/hir/lowering.rs:855]: nit: probably ought to a separate file (doesn't need to be a submodule IMO)

**Review by Manishearth — APPROVED** [review]
This looks pretty great! The pipeline model for handling lifetimes is pretty elegant :smile: 

There's a small bug, beyond that I think you should go ahead and merge early and maybe handle stuff in a followup if you need.

File: core/src/hir/lowering.rs
Comment (QnnOkabayashi) [reply core/src/hir/lowering.rs:902]: Different states have different meanings. For example, the `SelfParamLifetimeLowerer` type is only used on the self type, since the self type is special in elision inference. There's only one self type (or none), so it's nice to have functions the consume ownership of the type, proving at the type level it's only called once. Same with the second state, which has the potential to set the lifetime that elision in the output refers to, whereas the last state can only read that state. They all lower lifetimes, but they represent different stages in elision inference, so it's nice to have a state machine.

File: core/src/hir/lowering.rs
Comment (QnnOkabayashi) [reply core/src/hir/lowering.rs:1089]: That's a copy paste bug, the body came from `no_self_lifetime`, will fix this now. Should probably disable the `allow(unused_*)` lint in the hir module...

File: core/src/hir/lowering.rs
Comment (QnnOkabayashi) [reply core/src/hir/lowering.rs:855]: All of this is still specifically related to lowering, so I felt like this was appropriate. I could definitely move it to a separate file though.

I agree that this seriously needs docs, but I'd rather write them once I resolve the corner case I mentioned. The PR for that fix (and these docs) should be pretty quick once this PR is merged since I already know the fix and didn't want to bloat this PR too much.

**Pull request merged**
